security/oc_certs_generate.c

```diff
@@ -318,29 +360,164 @@ oc_certs_encode_roles(const oc_role_t *roles,
     roles = roles->next;
   }
 
-  *general_names = head;
+  *names = head;
   return count;
 }
 
+#else /* MBEDTLS_VERSION_NUMBER > 0x03010000 */
+
+/// @brief Deallocate a linked list of mbedtls_x509_san_list*
+static void
+certs_free_encoded_roles(mbedtls_x509_san_list *names)
+{
+  while (names != NULL) {
+    mbedtls_x509_san_list *next = names->next;
+    assert(names->node.type == MBEDTLS_X509_SAN_DIRECTORY_NAME);
+    mbedtls_free(names->node.san.directory_name.oid.p);
+    mbedtls_free(names->node.san.directory_name.val.p);
+    if (names->node.san.directory_name.next != NULL) {
+      mbedtls_asn1_free_named_data_list(&names->node.san.directory_name.next);
+    }
+    free(names);
+    names = next;
+  }
+}
+
+static mbedtls_x509_san_list *
+certs_encode_role(const oc_role_t *role)
+{
+  char roleid[MBEDTLS_X509_MAX_DN_NAME_SIZE] = { 0 };
+  if (!oc_certs_encode_role(role, roleid, OC_ARRAY_SIZE(roleid))) {
+    OC_ERR("error encoding roleid");
+    return NULL;
+  }
+  /* A RoleId is encoded in a GeneralName that is of type directoryName into
+   * the GeneralNames SEQUEENCE.
+   */
+  mbedtls_x509_san_list *name =
+    (mbedtls_x509_san_list *)calloc(1, sizeof(mbedtls_x509_san_list));
```

Comment:
calloc could be used only if OC_DYNAMIC_ALLOCATION is defined.

Reply:
the whole file is under `ifdef OC_DYNAMIC_ALLOCATION `